motionplan/tpSpaceRRT.go

```diff
@@ -569,64 +585,66 @@ func (mp *tpSpaceRRTMotionPlanner) extendMap(
 		}
 	}
 	treeNode := bestCand.treeNode // The node already in the tree to which we are parenting
-	newNode := bestCand.newNode   // The node we are adding because it was the best extending PTG
+	newNodes := bestCand.newNodes // The node we are adding because it was the best extending PTG
 
-	ptgNum := int(newNode.Q()[0].Value)
-	randAlpha := newNode.Q()[1].Value
-	randDist := newNode.Q()[2].Value
-
-	trajK, err := mp.tpFrame.PTGSolvers()[ptgNum].Trajectory(randAlpha, randDist)
-	if err != nil {
-		return nil, err
-	}
+	for _, newNode := range newNodes {
+		ptgNum := int(newNode.Q()[0].Value)
+		randAlpha := newNode.Q()[1].Value
+		randDist := newNode.Q()[2].Value
 
-	arcStartPose := treeNode.Pose()
-	if invert {
-		arcStartPose = spatialmath.PoseBetweenInverse(trajK[len(trajK)-1].Pose, arcStartPose)
-	}
-	lastDist := 0.
-	sinceLastNode := 0.
+		trajK, err := mp.tpFrame.PTGSolvers()[ptgNum].Trajectory(randAlpha, randDist)
+		if err != nil {
+			return nil, err
+		}
+		arcStartPose := treeNode.Pose()
+		if invert {
+			arcStartPose = spatialmath.PoseBetweenInverse(trajK[len(trajK)-1].Pose, arcStartPose)
+		}
+		lastDist := 0.
+		sinceLastNode := 0.
 
-	var trajState *ik.State
-	if mp.algOpts.addIntermediate {
-		for i := 0; i < len(trajK); i++ {
-			trajPt := trajK[i]
-			if invert {
-				trajPt = trajK[(len(trajK)-1)-i]
-			}
-			if ctx.Err() != nil {
-				return nil, ctx.Err()
-			}
-			trajState = &ik.State{Position: spatialmath.Compose(arcStartPose, trajPt.Pose)}
-			if mp.algOpts.pathdebug {
-				if !invert {
-					mp.logger.Debugf("$FWDTREE,%f,%f\n", trajState.Position.Point().X, trajState.Position.Point().Y)
-				} else {
-					mp.logger.Debugf("$REVTREE,%f,%f\n", trajState.Position.Point().X, trajState.Position.Point().Y)
+		var trajState *ik.State
+		if mp.algOpts.addIntermediate {
+			for i := 0; i < len(trajK); i++ {
+				trajPt := trajK[i]
+				if invert {
+					trajPt = trajK[(len(trajK)-1)-i]
 				}
-			}
-			sinceLastNode += (trajPt.Dist - lastDist)
-
-			// Optionally add sub-nodes along the way. Will make the final path a bit better
-			// Intermediate nodes currently disabled on the goal map because they do not invert nicely
-			if sinceLastNode > mp.algOpts.addNodeEvery && !invert {
-				// add the last node in trajectory
-				addedNode = &basicNode{
-					q:      referenceframe.FloatsToInputs([]float64{float64(ptgNum), randAlpha, trajPt.Dist}),
-					cost:   trajPt.Dist,
-					pose:   trajState.Position,
-					corner: false,
+				if ctx.Err() != nil {
+					return nil, ctx.Err()
 				}
-				rrt[addedNode] = treeNode
-				sinceLastNode = 0.
+				trajState = &ik.State{Position: spatialmath.Compose(arcStartPose, trajPt.Pose)}
+				if mp.algOpts.pathdebug {
+					if !invert {
```

Comment:
The particular spot you commented on is what allows us to visualize the fwd and reverse trees separately, so ironically is likely to be the longest-lived instance of an `invert` conditional even if we can abstract away the rest.

More broadly, since the invert of a trajectory requires radically different math than the forwards equivalent, there's not an easy way to do away with this.

The best abstraction I can think of would be dependent on RSDK-5101, where then we can allow the specification of the sub-pieces in any order, and push the invert math to there.

Reply:
Yeah that makes sense.  Was just wondering, I definitely don't think its high priority but a nicer abstraction would pay dividends in terms of maintenance if we can figure that out somehow

Reply:
Yeah, for sure. I think there are a few ways this algo can probably be simplified in that way; note that its linecount in currently nearly 2x that of cbirrt.